autonomi/src/client/high_level/vault/mod.rs

```diff
@@ -52,24 +77,86 @@ impl Client {
         secret_key: &VaultSecretKey,
     ) -> Result<(Bytes, VaultContentType), VaultError> {
         info!("Fetching and decrypting vault...");
-        let public_key = secret_key.public_key();
-        let pad = self.scratchpad_get_from_public_key(&public_key).await?;
+        let main_secret_key = MainSecretKey::new(secret_key.clone());
+        let public_key = main_secret_key
+            .derive_key(&DerivationIndex::from_bytes(VAULT_HEAD_DERIVATION_INDEX))
+            .public_key();
+
+        let mut cur_graph_entry_addr = GraphEntryAddress::from_owner(public_key.into());
+        let mut decrypted_full_text = vec![];
+        let mut content_type = 0;
+        let mut has_end_reached = false;
+
+        while !has_end_reached {
+            let graph_entry = self.graph_entry_get(cur_graph_entry_addr).await?;
+
+            // The first descendant is reserved for `expand GraphEntry`.
+            match graph_entry.descendants.split_first() {
+                Some((&(first, _), rest)) => {
+                    cur_graph_entry_addr = GraphEntryAddress::from_owner(first);
+                    let scratchpad_addresses = rest.to_vec();
+
+                    let (decrypt_data, cur_content_type, is_end_reached) = self
+                        .fetch_scratchpads_of_one_graph_entry_and_decrypt(
+                            &main_secret_key,
+                            scratchpad_addresses,
+                        )
+                        .await?;
+                    decrypted_full_text.push(decrypt_data);
+                    content_type = cur_content_type;
+                    has_end_reached = is_end_reached;
+                }
+                None => {
+                    let msg = format!(
+                        "Vault's GraphEntry at {cur_graph_entry_addr:?} only has {} descendants.",
+                        graph_entry.descendants.len()
+                    );
+                    return Err(VaultError::VaultNotEnoughGraphDescendants(msg));
+                }
+            }
+        }
 
-        let data = pad.decrypt_data(secret_key)?;
         debug!("vault data is successfully fetched and decrypted");
-        Ok((data, pad.data_encoding()))
+        Ok((Bytes::from(decrypted_full_text.concat()), content_type))
     }
 
     /// Get the cost of creating a new vault
-    pub async fn vault_cost(&self, owner: &VaultSecretKey) -> Result<AttoTokens, CostError> {
+    /// A quick estimation of cost:
+    ///   num_of_graph_entry * graph_entry_cost + num_of_scratchpad * scratchpad_cost
+    pub async fn vault_cost(
+        &self,
+        owner: &VaultSecretKey,
+        max_size: u64,
+    ) -> Result<AttoTokens, VaultError> {
```

Comment:
Here please return a `CostError` if possible. If possible this will allow to remove `CostError` from the `VaultError`